### ISSUE TITLE
Implement run containers

### ISSUE THREAD

**Issue by lemire** [RoaringBitmap/roaring#37]
For better compression when there are long runs of consecutive values, the Java version has run containers (as of version 0.5). They should be implemented :

https://github.com/lemire/RoaringBitmap/blob/master/src/main/java/org/roaringbitmap/RunContainer.java

These can be created by the "runOptimize" function...

https://github.com/lemire/RoaringBitmap/blob/master/src/main/java/org/roaringbitmap/RoaringBitmap.java#L1286

These containers can improve the compression ratio in some cases.

**Comment by lemire** [comment]
This is now used in Apache Spark and Druid with good results.

**Comment by lemire** [comment]
c.c. @brentp

**Comment by brentp** [comment]
so then, for every method, it needs to implement, e.g. 
```
bitmap.intersects(bitmap)
array.intersects(bitmap)
array.intersects(rle)
rle.intersects(bitmap)
```

**Comment by lemire** [comment]
@brentp 

Sadly, yes, it is a pain. On the plus side, there is a complete Java implementation that has been tuned and debugged. So copy-and-paste buys a lot of progress. Believe me, the first implementation (in Java) is surely orders of magnitude more troublesome than a simple port of the idea.

The benefits can be huge however in practice.

**Comment by brentp** [comment]
the benefits are in space? or time? must be space. I guess there could be a method rle.convert() that returns either a bitmap or an array depending on it's own density?

**Comment by lemire** [comment]
There are important gains in both space and time.

Here is the preprint :

http://roaringbitmap.org/RunRoaringBitmap.pdf

It is really very good. I was surprised at how well it did. Better yet: the good results were preserved when it went into production.

**Comment by brentp** [comment]
so the rle is single uint16 array (slice in go) with alternating values of [num_i, runlen_i-1, num_j, runlen_j-1, ...] 
is there an advantage of that over 2 separate arrays?

**Comment by lemire** [comment]
@brentp 

We interleave the data because it is always used together. Otherwise, the code is unnecessarily  complex and it is likely that there are more cache misses. Think about a binary search to check whether a value is contained. With two arrays, each access hits two cache lines. You can almost be sure to have twice as many cache misses.

Separating the data is indeed the correct default strategy when the data is commonly *not* used together. Then you have less data to go through and possibly higher speed. I don't think it is one of these cases.

**Comment by brentp** [comment]
so from the paper, it looks like Roaring without runs is almost always at least as fast as roaring+runs except for entire unions where roraing+runs is much faster.
In table III, it'd be interesting to know how much faster is Roaring+Run in IIIe than it is in IIId

**Comment by lemire** [comment]
> so from the paper, it looks like Roaring without runs is almost always at least as fast as roaring+runs except for entire unions where roraing+runs is much faster.

Yes, but consider Table Ic. In some instances, Roaring+Run compresses the data much better, while remaining really fast.

> In table III, it'd be interesting to know how much faster is Roaring+Run in IIIe than it is in IIId

It is not presented directly but it can be inferred because Roaring+Run naive appears in IIIe.

**Comment by brentp** [comment]
got it. thanks.

**Comment by brentp** [comment]
I started having a look at this, but might just wait to wrap the C lib. :)
I did find that since the memory layout is under our control in go (as in C). It's nicer to have:

```go
type pair struct {
	start uint16
	run   uint16
}
type rle []pair
```

than 

```
type rle []uint16
```

becuase with the former, we can just index into `rle` as normal and don't have to worry about multiplying by 2 in the search.

Below is as far as I got:

```go
import "fmt"

type pair struct {
	start uint16
	run   uint16
}

func (p pair) end() uint16 {
	return p.start + p.run
}

func (p pair) String() string {
	return fmt.Sprintf("pair{start: %d, run: %d}", p.start, p.run)
}

type rle []pair

var _ container = &rle{}

func (r *rle) and(b *bitmap) *bitmap {
	shared := make([]uint16, 0, 4)
	for _, v := range r.shorts() {
		if b.get(v) {
			shared = append(shared, v)
		}
	}
	out := &bitmap{}
	out.set(shared...)
	return out
}

func (r *rle) xor(b *bitmap) *bitmap {
	out := &bitmap{}
	out.bitmap = b.bitmap
	for _, v := range r.shorts() {
		if out.get(v) {
			out.clear(v)
		} else {
			out.set(v)
		}
	}
	return out
}

func (r *rle) or(b *bitmap) *bitmap {
	out := &bitmap{}
	out.bitmap = b.bitmap
	for _, v := range r.shorts() {
		out.set(v)
	}
	return out
}

func NewRLE(vals ...uint16) *rle {
	m := make([]pair, 0, 5)
	if len(vals) > 0 {
		last := vals[0]
		first := last
		run := uint16(0)
		for _, v := range vals[1:] {
			if v-1 == last {
				run++
			} else {
				m = append(m, pair{first, run})
				first = v
			}
			last = v
		}
		m = append(m, pair{first, run})
	}
	r := rle(m)
	return &r
}

func (r *rle) delete(i int) {
	(*r) = append((*r)[:i], (*r)[i+1:]...)
}

func (r *rle) set(vals ...uint16) {
	for _, val := range vals {
		idx, in := r.search(val)
		if in {
			continue
		}
		if len(*r) == 0 {
			(*r) = append(*r, pair{val, 0})
			continue
		}
		if idx >= 0 {
			(*r)[idx].run += 1
			// check if we extended to the left
			// in which case, we decrement the start.
			if (*r)[idx].start > val {
				(*r)[idx].start -= 1
				// check if we need to merge blocks.
				if idx > 0 && (*r)[idx].start == (*r)[idx-1].end()+1 {
					(*r)[idx-1].run += (*r)[idx].run + 1
					r.delete(idx)
				}
			} else {
				// check for merge.
				if len(*r) <= idx+1 {
					continue
				}
				end := (*r)[idx].end()
				if end+1 == (*r)[idx+1].start {
					(*r)[idx].run += (*r)[idx+1].run + 1
					r.delete(idx + 1)
				}
			}
		} else {
			idx = (-idx) - 1
			(*r) = append((*r)[:idx], append([]pair{pair{val, 0}}, (*r)[idx:]...)...)
		}
	}

}

func (r *rle) get(key uint16) bool {
	_, in := r.search(key)
	return in
}

func (r *rle) search(key uint16) (int, bool) {
	low, high := 0, len(*r)-1
	for low+16 <= high {
		mid := int(uint32(low+high) >> 1)
		midPair := (*r)[mid]
		// check: <= ? since run is 0-based
		if midPair.start+midPair.run < key {
			low = mid + 1
		} else if midPair.start > key {
			high = mid - 1
		} else {
			return mid, true
		}
	}
	for ; low <= high; low++ {
		block := (*r)[low]
		// key is above current block
		if key > block.start+block.run+1 {
			continue
		}
		// key is below current block
		if block.start > 0 && key < block.start-1 {
			continue
		}
		// so now we know the key can be a member of the current block.
		// need to check if it already exists in the block.
		if key < block.start+block.run+1 && key >= block.start {
			return low, true
		}
		// or if it is adjacent to the current block
		return low, false
	}
	return -(low + 1), false

}

func (r *rle) cardinality() int {
	n := 0
	for _, p := range *r {
		n += int(p.run) + 1
	}
	return n
}

func (r *rle) shorts() []uint16 {
	s := make([]uint16, r.cardinality())
	j := 0
	for _, p := range *r {
		for i := 0; i < int(p.run)+1; i++ {
			s[j] = p.start + uint16(i)
			j++
		}
	}
	return s
}
```

**Comment by lemire** [comment]
@brentp 

I think we need to integrate your code. Want to issue a PR? Alternatively, we can grant you access to the repo... would be wise to start out in a branch until we have integration and unit tests, of course. In any case, we would want to add you as a contributor (see README file). How do you want to go about it?

> wait to wrap the C lib

There are a couple of C versions of Roaring, but I don't expect wrapping it in Go is going to be practical. Last time I checked, any function call from Go to C had an overhead of hundreds of cycles at the very least. Even wrapping Go's own assembly is not all that practical because Google disables the assembly path in its cloud hosting solutions.

I am not a "purist" in any sense, but it makes a lot of practical sense, I think, to generate pure Go solutions.

That's quite aside for production issues... where running Go code that has C dependencies not likely to be automagically resolve make deployment fragile.



>  we can just index into rle as normal and don't have to worry about multiplying by 2 in the search.

Oh!!! This is so obvious now that you point it out!!! Oh the pain that you are saving us by pointing this out.

To be pedantic... Note that I am not sure the C langage guarantees that the struct will use 4 bytes, though I suppose it must be true on x86 processors...  but it looks like Go might enforce, though I could not find it stated without ambiguity.

**Comment by brentp** [comment]
I figured a go wrapper would be practical even with the call overhead since the operations will be on a full container, not on each (uint16) element in a container.

**Comment by brentp** [comment]
I think you already added me to contributors. You're welcome to incorporate this code as-is, though it will need some modifications as I had my own minimal bitmap container that I'm not sure is compatible with the one from this repo.

**Comment by lemire** [comment]
> You're welcome to incorporate this code as-is, though it will need some modifications 

Ok. Let us leave it there for now until I or someone else can make good use of it.

> I think you already added me to contributors.

You can't blame me for being overeager on such things... ;-)

> I figured a go wrapper would be practical even with the call overhead since the operations will be on a full container, not on each (uint16) element in a container.

For unions and intersections... sure... for contains/add/iteration... I think that the C-to-go overhead will be too large.

**Comment by lemire** [comment]
It has been implemented in C...
https://github.com/RoaringBitmap/CRoaring/blob/master/include/roaring/containers/run.h

**Comment by glycerine** [comment]
I did an initial Go RunContainer for uint32, with the idea that uint16 would be trivial copy and paste once it is complete. It's available in pull request #70. It probably needs additions to integrate it with Bitmap, but there's alot of functionality there, and it is well tested.

Both the C and the java binary search implementations are subject to a classic integer overflow bug in binary search; e.g. at
https://github.com/RoaringBitmap/CRoaring/blob/master/include/roaring/containers/run.h#L91
instead of
`int32_t middleIndex = (low + high) >> 1;`
prefer:
`int32_t middleIndex = low + ((high-low) >> 1);`

**Comment by glycerine** [comment]
an update and a question:

https://github.com/RoaringBitmap/roaring/pull/70 now has uint16 and uint32 run-container backends. The container interface methods are stubbed out but not implemented.

Question: currently I am using [first, last] closed intervals. While my intitial implementation used [first, excluded_end) half-open intervals, I realized that this didn't support storing the maximum 2^16-1 (for uint16) or 2^32-1 (for uint32) value in the container. Hence I switched.

I notice however that the C implementation uses start + runlength. This seems to me to create partitioning problems For example, a run be set that crosses the 2^16 container boundaries: [start: 2^16-2, length: 4] would cross into the next 2^16 sized bucket. Is this a concern?  Maybe it doesn't arise in practice. But since one wouldn't want to cross these bucket boundaries (I have to suppose), then there is no advantage of the start+length over the [start, last] representation is there?  Generally I think of [start, endx) as easiest to maintain, then [start, last]; hence I am preferring [start, last] for ease of maintenance and for access to the full 2^N-1 bitspace.  

I'm happy to hear arguments from another viewpoint; please advise if there is a compelling reason to use run-length instead of last-value to describe the spans.

**Comment by jaten-saucelabs** [comment]
Update:
as of b4b9700827e9b2041cc95312f6f27a5f4756929a, the #70 PR is almost complete. 

runContainer16 is missing only 3 methods of the container interface:
xor, lazyIOR, and lazyOR.  These shouldn't be difficult, although I'm not exactly clear on
the semantics of the lazy methods.

Also the container interface was somewhat inconsistent, with methods like add and remove sometimes acting inplace and sometimes not; I've tried to address this, for example, by splitting into iadd and iaddReturnMinimized. The convention for the ReturnMinimized suffixed methods is that they may change container type. This simplified certain aspects of testing.

@lemire I may not have time -- at least for a while -- to work on this again in the short term, so I invite anyone is able to finish this up. I wrote tests as I went, so coverage of the central file, rle16.go is very good; currently at 94%.

tests are still needed for the following. I assume that existing arraycontainer_test.go or bitmapcontainer tests can be readily adapted.

* fillLeastSignificant16bits()
* getShortIterator()
* getSizeInBytes() 
* iaddRange()
* iremoveRange()
* not(firstOfRange, lastOfRange int) container
* equals()
* intersects

these three need both -- i.e. both tests and implementation. Implementation is stubbed out in rlei.go:
* lazyIOR()
* lazyOR()
* xor()

**Comment by lemire** [comment]
Reposting here a comment made regarding serialization:

I think we want to keep the existing ``writeTo`` and ``readFrom`` methods.  We have an efficient  format that is supported across a wide range of libraries, and is being used in important production systems, and whose performance and data size has been documented and well tested. We have that ``gocroaring`` (https://github.com/RoaringBitmap/gocroaring), ``RoaringBitmap`` (https://github.com/RoaringBitmap/RoaringBitmap), ``CRoaring`` (https://github.com/RoaringBitmap/CRoaring), ``CRoaring.Net`` (https://github.com/RogueException/CRoaring.Net), ``croaring-rs`` (https://github.com/saulius/croaring-rs) and ``PyRoaringBitMap`` (https://github.com/Ezibenroc/PyRoaringBitMap) all have the same serialization format. So  I suggest we keep ``writeTo``, ``serializedSizeInBytes``, ``GetSerializedSizeInBytes`` and ``readFrom`` methods as they are, and add `writeToMessagePack``, ``serializedMessagePackSizeInBytes``, ``GetSerializedMessagePackSizeInBytes`` and ``readFromMessagePack`` if you feel  that MessagePack is an important contribution. If adding support to run containers to ``writeTo`` and ``readFrom`` is an issue,  I will handle it, it is not very difficult.

**Comment by lemire** [comment]
@glycerine 

> I notice however that the C implementation uses start + runlength. This seems to me to create partitioning problems For example, a run be set that crosses the 2^16 container boundaries: [start: 2^16-2, length: 4] would cross into the next 2^16 sized bucket. Is this a concern? 

No. Not that I am aware.

> But since one wouldn't want to cross these bucket boundaries (I have to suppose), then there is no advantage of the start+length over the [start, last] representation is there? 

Right. Your implementation is fine.

> Generally I think of [start, endx) as easiest to maintain, then [start, last]; hence I am preferring [start, last] for ease of maintenance and for access to the full 2^N-1 bitspace.

I think that [start, last] is fine.

**Comment by glycerine** [comment]
> So I suggest we keep writeTo, serializedSizeInBytes, GetSerializedSizeInBytes and readFrom methods as they are, and add writeToMessagePack, serializedMessagePackSizeInBytes, GetSerializedMessagePackSizeInBytes and readFromMessagePack if you feel that MessagePack is an important contribution. If adding support to run containers to writeTo and readFrom is an issue, I will handle it, it is not very difficult.

Sounds good. Since it's already done, let's keep the msgpack2/snappy for comparison. 

Also: good context. I wasn't aware that there was already a documented format for run-containers (or for the other types). If you point me at the serialization format docs, I can probably do a quick determination of whether I can tackle it. 

One thing that puzzled me in the current master branch was that the container size implicitly indicated the type on disk. Hopefully there is an actual type indicator byte present in the current format (ala contype in the msgpack serializer), so that if the user forgets to optimize (or doesn't want to), they don't end up confusing the deserializer.

**Comment by lemire** [comment]
@glycerine The format specification can be found there: 

https://github.com/RoaringBitmap/RoaringFormatSpec

The current go implementation fully support (efficiently) a subset of the specification.

**Comment by lemire** [comment]
Fixed in release 0.3.0 by @glycerine 

https://github.com/RoaringBitmap/roaring/releases/tag/v0.3.0
